### PR TITLE
Share podman sock bindings with other WSL distros

### LINKED ISSUE #15190: Allow to automatically set podman connection for other WSL2 instances
For WSL2 a new instance is created and a connection is added for interaction from the host using the podman-remote client, however people might like to use podman from other WSL installed distros on the same machine. For this, a new connection would need to be added to those instances. Enabling this in an easy way using a command?

Connections are composed as follows:

```
podman-machine-default       ssh://user@localhost:59971/run/user/1000/podman/podman.sock  C:\Users\gbraad\.ssh\podman-machine-default  false
podman-machine-default-root  ssh://root@localhost:59971/run/podman/podman.sock            C:\Users\gbraad\.ssh\podman-machine-default  false
```

the port, eg `59971` in this example is randomly generated.

ref: #15107

### PR DESCRIPTION
**Pull request by n1hility**
Resolves #15190

Registers a rootless and rootful socket underneath /mnt/wsl/podman-sockets/[machine name]/ This allows podman remote clients on other Linux distributions to access podman.

This also registers the podman root socket under the wheel group, to allow for rootful linking against /var/run/docker.sock, a use case expected by some clients and APIs. While this is not recommended practice on a Linux host, a WSL guest is user-isolated and already enables escalation trivially.

```release-note
Registers shared socket bindings on Windows, to allow other WSL distributions easy remote access
```

### CODE REVIEW

**Comment by n1hility** [comment]
PTAL @containers/podman-maintainers

**Review by vrothberg — COMMENTED** [review]
Code LGTM but I have no means to test it at the moment

File: pkg/machine/wsl/machine.go
Comment (Luap99) [inline pkg/machine/wsl/machine.go:175]: why does this use fstab at all? By the time fstab tries to mount it the files will not be there anyway so it will fail and only be useful for the mount command in the unit.
But then you might as well just type the full mount command into the unit and skip the fstab part.


Also you can create .mount units with systemd then you could just fire up with the depedncy management of systemd at the right time.

File: pkg/machine/wsl/machine.go
Comment (Luap99) [inline pkg/machine/wsl/machine.go:166]: I think this should be type=oneshot?

File: pkg/machine/wsl/machine.go
Comment (Luap99) [inline pkg/machine/wsl/machine.go:151]: also type=oneshot?

File: pkg/machine/wsl/machine.go
Comment (Luap99) [inline pkg/machine/wsl/machine.go:182]: can we somehow make the paths into constants so that we can match them with configureBindMounts() below.
Otherwise someone might forget to change it.

Second, why to we use `[USER]` here compared to the normal sprintf() style? This seems rather inconstant and hard to follow.

**Review by Luap99 — COMMENTED** [review]
The whole disconnect between the constant and the use worries me, this is hard to follow.
What uses printf() stlye what uses the weird `[USER]` syntax. AT the caller it is impossible to know I would have to look this up every single time.

The me a much easier to understand approach would be to create functions that return the string, i.e. create function `getBindMountUserService(dist string) string` then I automatically know what argument to use.

File: pkg/machine/wsl/machine.go
Comment (n1hility) [reply pkg/machine/wsl/machine.go:175]: @Luap99 it needs fstab because one of the services has a dependency on a `systemd --user` process, so has to run as non-root, and root is normally required to issue mounts unless there is a pre-declared fstab entry (user option). It's also marked as noauto to keep it from being picked up by anything other than this mount.

The only reason this wasn't done as mount unites was just to keep everything as just one unit file [per root/user] (in addition to the mount it also needs to precreate the tmpfs directory structure).  Unfortunately this has to be duplicated in both units since systemd does not support dependencies that cross user/root.

File: pkg/machine/wsl/machine.go
Comment (n1hility) [reply pkg/machine/wsl/machine.go:166]: Yes you are right this should be oneshot.

File: pkg/machine/wsl/machine.go
Comment (n1hility) [reply pkg/machine/wsl/machine.go:182]: Sure I can improve that

**Comment by n1hility** [comment]
> The whole disconnect between the constant and the use worries me, this is hard to follow. What uses printf() stlye what uses the weird `[USER]` syntax. AT the caller it is impossible to know I would have to look this up every single time.
> 
> The me a much easier to understand approach would be to create functions that return the string, i.e. create function `getBindMountUserService(dist string) string` then I automatically know what argument to use.

This file is definitely due some cleanup. There was originally a very small number of constants and a very limited number of replacements, but it's grown over time to be a lot. In a future PR I want to refactor this a bit and split things up so it's easier to follow. For this PR I will improve the aspects related to this change though

File: pkg/machine/wsl/machine.go
Comment (Luap99) [reply pkg/machine/wsl/machine.go:175]: Ah right, that makes sense then.

**Comment by Luap99** [comment]
Yes, I am not blocking on that I am only adding these comments because it costs me time to understand/review the change because I constantly need jump around to see what arg is given, etc...
If you like to clean up this in the future this would be much appreciated.

**Comment by mheon** [comment]
LGTM once comments from @Luap99 are addressed

**Comment by n1hility** [comment]
> Yes, I am not blocking on that I am only adding these comments because it costs me time to understand/review the change because I constantly need jump around to see what arg is given, etc...
> If you like to clean up this in the future this would be much appreciated.

Yeah agree, the code should aim to minimize review time. I did some improvements for this change, will  definitely look at a more complete follow-up in the future.

**Comment by n1hility** [comment]
It looks like the bud and centos stream failures are unrelated. 

All feedback should be addressed

File: pkg/machine/wsl/machine.go
Comment (TomSweeneyRedHat) [inline pkg/machine/wsl/machine.go:155]: is `777` realy correct?  Seems way to open to me, but....

Comment (n1hility) [reply pkg/machine/wsl/machine.go:155]: Yeah it’s certainly an odd setup. So in this case /mnt/wsl (outside of our control is 777) since it’s a shaded space (sorta like /tmp) where all WSl distros cross (different Linux distributions, each running in a linux namespace). This subdir underneath just mirrors that permission scheme, and is also necessary since there is both a user systemd service and a root systemd service writing into the same directory.

**Review by Luap99 — COMMENTED** [review]
LGTM

**Review by vrothberg — APPROVED** [review]
/lgtm

**Pull request merged**
